bundles/org.openhab.binding.luxom/src/main/java/org/openhab/binding/luxom/internal/handler/LuxomBridgeHandler.java

```diff
@@ -0,0 +1,346 @@
+    }
+
+    private synchronized void reconnect(boolean timeout) {
+        if (timeout) {
+            logger.error("Keepalive timeout, attempting to reconnect to the bridge");
+        } else {
+            logger.debug("Connection problem, attempting to reconnect to the bridge");
+        }
+
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.DUTY_CYCLE);
+        disconnect();
+        connect();
+    }
+
+    public void sendCommands(List<CommandExecutionSpecification> commands) {
+        this.sendQueue.add(commands);
+    }
+
+    @Nullable
+    private LuxomThingHandler findThingHandler(@Nullable String address) {
+        for (Thing thing : getThing().getThings()) {
+            if (thing.getHandler() instanceof LuxomThingHandler) {
+                LuxomThingHandler handler = (LuxomThingHandler) thing.getHandler();
+
+                try {
+                    if (handler != null && handler.getAddress().equals(address)) {
+                        return handler;
+                    }
+                } catch (IllegalStateException e) {
+                    logger.trace("Handler for id {} not initialized", address);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * needed with fast reconnect to update status of things
+     */
+    public void forceRefreshThings() {
+        for (Thing thing : getThing().getThings()) {
+            if (thing.getHandler() instanceof LuxomThingHandler) {
+                LuxomThingHandler handler = (LuxomThingHandler) thing.getHandler();
+                handler.ping();
+            }
+        }
+    }
+
+    private void sendHeartBeat() {
+        logger.trace("Sending heartbeat");
+        // Reconnect if no response is received within KEEPALIVE_TIMEOUT_SECONDS.
+        heartBeatTimeoutTask = scheduler.schedule(() -> this.reconnect(true), HEARTBEAT_ACK_TIMEOUT_SECONDS,
+                TimeUnit.SECONDS);
+        sendCommands(Collections.singletonList(new CommandExecutionSpecification(LuxomAction.HEARTBEAT.getCommand())));
+    }
+
+    @Override
+    public void thingUpdated(Thing thing) {
+        LuxomBridgeConfig newConfig = thing.getConfiguration().as(LuxomBridgeConfig.class);
+        boolean validConfig = validConfiguration(newConfig);
+        boolean needsReconnect = validConfig && this.config != null && !this.config.sameConnectionParameters(newConfig);
+
+        if (!validConfig || needsReconnect) {
+            dispose();
+        }
+
+        this.thing = thing;
+        this.config = newConfig;
+
+        if (needsReconnect) {
+            initialize();
+        }
+    }
+
+    public void handleCommunicationError(IOException e) {
+        logger.error("Communication error while reading, will try to reconnect. Error: {}", e.getMessage());
+        reconnect();
+    }
+
+    @Override
+    public void dispose() {
+        disconnect();
+    }
+
+    public void handleIncomingLuxomMessage(String luxomMessage) throws IOException {
```

Comment:
As far as I can see this is called from a thread - is it thread-safe?

Reply:
As far as I can see: yes. 

The thread is the loop of the Buffered reader, it will not execute in parallel.